## Comprendre notre DB excel

In [42]:
import pandas as pd

In [43]:
players = pd.read_excel('../inputs/regular NBA.xlsx', sheet_name='Données NBA', header=1)
teams = pd.read_excel('../inputs/regular NBA.xlsx', sheet_name='Equipe')
descriptions = pd.read_excel('../inputs/regular NBA.xlsx', sheet_name='Dictionnaire des données')

In [44]:
players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 53 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Player       569 non-null    object 
 1   Team         569 non-null    object 
 2   Age          569 non-null    int64  
 3   GP           569 non-null    int64  
 4   W            569 non-null    int64  
 5   L            569 non-null    int64  
 6   Min          569 non-null    float64
 7   PTS          569 non-null    int64  
 8   FGM          569 non-null    int64  
 9   FGA          569 non-null    int64  
 10  FG%          569 non-null    float64
 11  15:00:00     569 non-null    int64  
 12  3PA          569 non-null    int64  
 13  3P%          569 non-null    float64
 14  FTM          569 non-null    int64  
 15  FTA          569 non-null    int64  
 16  FT%          569 non-null    float64
 17  OREB         569 non-null    int64  
 18  DREB         569 non-null    int64  
 19  REB     

In [45]:
players.head()

Player Team  Age  GP   W   L   Min   PTS  FGM   FGA  ...  \
0  Shai Gilgeous-Alexander  OKC   26  76  63  13  34.2  2485  859  1657  ...   
1          Anthony Edwards  MIN   23  79  48  31  36.3  2180  719  1612  ...   
2             Nikola Jokić  DEN   30  70  46  24  36.7  2072  784  1365  ...   
3    Giannis Antetokounmpo  MIL   30  67  40  27  34.2  2037  791  1320  ...   
4             Jayson Tatum  BOS   27  72  53  19  36.4  1930  662  1462  ...   

    PIE  POSS  Unnamed: 45  Unnamed: 46  Unnamed: 47  Unnamed: 48  \
0  19.9  5514          NaN          NaN          NaN          NaN   
1  14.0  5883          NaN          NaN          NaN          NaN   
2  20.6  5471          NaN          NaN          NaN          NaN   
3  21.0  4772          NaN          NaN          NaN          NaN   
4  15.8  5368          NaN          NaN          NaN          NaN   

   Unnamed: 49  Unnamed: 50  Unnamed: 51  Unnamed: 52  
0          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN  

[5 rows x 53 columns]

In [46]:
players.dropna(inplace=True, axis=1)

In [47]:
players.describe()

Age          GP           W           L         Min  \
count  569.000000  569.000000  569.000000  569.000000  569.000000   
mean    26.152900   46.231986   23.235501   22.996485   19.429174   
std      4.217189   24.357135   15.596499   13.985404    9.539416   
min     19.000000    1.000000    0.000000    0.000000    0.900000   
25%     23.000000   25.000000   10.000000   11.000000   11.800000   
50%     25.000000   50.000000   21.000000   23.000000   19.500000   
75%     28.000000   68.000000   36.000000   32.000000   27.200000   
max     40.000000   82.000000   64.000000   64.000000   37.700000   

               PTS         FGM          FGA         FG%    15:00:00  ...  \
count   569.000000  569.000000   569.000000  569.000000  569.000000  ...   
mean    492.117750  180.304042   385.899824   44.610896   58.534271  ...   
std     473.851423  170.954232   357.177638   11.463917   63.585470  ...   
min       0.000000    0.000000     0.000000    0.000000    0.000000  ...   
25%     107.000000   41.000000    91.000000   40.600000    6.000000  ...   
50%     356.000000  133.000000   284.000000   44.800000   32.000000  ...   
75%     731.000000  271.000000   581.000000   49.600000   98.000000  ...   
max    2485.000000  859.000000  1657.000000  100.000000  324.000000  ...   

            OREB%       DREB%        REB%    TO RATIO        EFG%         TS%  \
count  569.000000  569.000000  569.000000  569.000000  569.000000  569.000000   
mean     4.820387   13.125308    8.944288    9.982250   51.719156   54.554833   
std      3.596625    5.585286    4.016021    4.184266   12.181426   11.453414   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      2.400000    9.100000    6.000000    7.900000   48.500000   51.600000   
50%      3.900000   11.800000    8.100000    9.700000   53.300000   56.300000   
75%      6.400000   16.400000   11.200000   11.800000   57.500000   60.000000   
max     25.000000   44.400000   22.100000   44.400000  125.000000  125.000000   

             USG%        PACE         PIE         POSS  
count  569.000000  569.000000  569.000000   569.000000  
mean    17.836907  101.637821    8.870650  2190.991213  
std      5.577496    3.695111    3.955234  1685.126979  
min      0.000000   88.670000   -8.700000     6.000000  
25%     14.100000   99.430000    6.700000   655.000000  
50%     16.900000  101.260000    8.700000  1903.000000  
75%     21.000000  103.180000   11.000000  3606.000000  
max     45.500000  130.810000   40.000000  6200.000000  

[8 rows x 43 columns]

In [48]:
sum(players.duplicated(subset='Player'))

0

In [49]:
players.shape

(569, 45)

In [50]:
# Passons aux équipes
teams.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 2 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Code                     30 non-null     object
 1   Nom complet de l'équipe  30 non-null     object
dtypes: object(2)
memory usage: 612.0+ bytes


In [51]:
teams.head()

Code Nom complet de l'équipe
0  ATL           Atlanta Hawks
1  BKN           Brooklyn Nets
2  BOS          Boston Celtics
3  CHA       Charlotte Hornets
4  CHI           Chicago Bulls

In [52]:
teams.shape

(30, 2)

In [53]:
# Passons aux dernier fichiers
descriptions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45 entries, 0 to 44
Data columns (total 2 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Dictionnaire des données  45 non-null     object
 1   Unnamed: 1                45 non-null     object
dtypes: object(2)
memory usage: 852.0+ bytes


In [54]:
descriptions.head()

Dictionnaire des données                                         Unnamed: 1
0                   Player                                      Nom du joueur
1                     Team                Équipe du joueur (code à 3 lettres)
2                      Age                                      Âge du joueur
3                       GP              Nombre de matchs joués (Games Played)
4                        W  Nombre de victoires de l'équipe lors des match...

In [55]:
descriptions.columns = ["Data", "Descriptions"]

In [56]:
descriptions.head()

Data                                       Descriptions
0  Player                                      Nom du joueur
1    Team                Équipe du joueur (code à 3 lettres)
2     Age                                      Âge du joueur
3      GP              Nombre de matchs joués (Games Played)
4       W  Nombre de victoires de l'équipe lors des match...

In [57]:
descriptions.shape

(45, 2)

###  Synthèse des données NBA

Ce projet utilise un fichier Excel composé de **trois feuilles complémentaires** permettant d’analyser les performances des joueurs NBA.

---

#### 1. `players` — Données principales des joueurs  
- **569 lignes × 45 colonnes**  
- **1 ligne = 1 joueur**  
- Contient les statistiques individuelles et avancées :  
  - Informations générales : `Player`, `Team`, `Age`  
  - Statistiques de base : `GP`, `W`, `L`, `Min`, `PTS`, `FGM`, `FGA`, `FG%`  
  - Statistiques avancées : `OREB%`, `DREB%`, `REB%`, `EFG%`, `TS%`, `USG%`, `PACE`, `PIE`, `POSS`  
- C’est la table principale utilisée pour l’EDA.

---

#### 2. `teams` — Référentiel des équipes  
- **30 lignes × 2 colonnes**  
- Colonnes :  
  - `Code` : code équipe (ex : ATL, BOS, LAL)  
  - `Nom complet de l'équipe` : nom officiel de la franchise  
- Permet d’enrichir `players` via une jointure sur le code équipe.

---

#### 3. `descriptions` — Dictionnaire de données  
- **45 lignes × 2 colonnes**  
- Colonnes :  
  - `Data` : nom de la variable  
  - `Descriptions` : définition de la variable  
- Correspond exactement aux 45 colonnes de `players`.  
- Sert de documentation pour comprendre chaque statistique.

---

#### Relations entre les tables  
- `players.Team` ↔ `teams.Code`  
  -> permet d’ajouter le nom complet des équipes  
- `players.columns` ↔ `descriptions.Data`  
  -> permet d’associer une définition à chaque variable

---

#### Structure de notre base de données
- `teams.Code` -> clé primaire
- `players.Team` -> clé étrangère
- `descriptions.Data` -> clé primaire

In [58]:
players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 45 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Player     569 non-null    object 
 1   Team       569 non-null    object 
 2   Age        569 non-null    int64  
 3   GP         569 non-null    int64  
 4   W          569 non-null    int64  
 5   L          569 non-null    int64  
 6   Min        569 non-null    float64
 7   PTS        569 non-null    int64  
 8   FGM        569 non-null    int64  
 9   FGA        569 non-null    int64  
 10  FG%        569 non-null    float64
 11  15:00:00   569 non-null    int64  
 12  3PA        569 non-null    int64  
 13  3P%        569 non-null    float64
 14  FTM        569 non-null    int64  
 15  FTA        569 non-null    int64  
 16  FT%        569 non-null    float64
 17  OREB       569 non-null    int64  
 18  DREB       569 non-null    int64  
 19  REB        569 non-null    int64  
 20  AST       

In [59]:
players.columns = players.columns.astype(str)
descriptions.columns = descriptions.columns.astype(str)
teams.columns = teams.columns.astype(str)

In [60]:
# comme nous avons beaucoups de colonne dans players (nous allons essayé de gérer la creation de la base automatiquement)
from sqlalchemy import Column, Integer, Float, String

# Nous allons écrire un scripts qui lit notre DataFrame
# détecte le type de chaque colonne et génère automatiquement le code SQLAlchemy correspondant

def sqlalchemy_type(dtype):
    if "int" in str(dtype):
        return "Integer"
    if "float" in str(dtype):
        return "Float"
    return "String"

print("class Player(Base):")
print("    __tablename__ = 'players'")
print("    id = Column(Integer, primary_key=True, autoincrement=True)")

for col, dtype in players.dtypes.items():
    col_name = col.replace(" ", "_").replace("%", "pct").replace("+/-", "plus_minus").replace(":", "_")
    sql_type = sqlalchemy_type(dtype)
    print(f"    {col_name} = Column({sql_type})")

class Player(Base):
    __tablename__ = 'players'
    id = Column(Integer, primary_key=True, autoincrement=True)
    Player = Column(String)
    Team = Column(String)
    Age = Column(Integer)
    GP = Column(Integer)
    W = Column(Integer)
    L = Column(Integer)
    Min = Column(Float)
    PTS = Column(Integer)
    FGM = Column(Integer)
    FGA = Column(Integer)
    FGpct = Column(Float)
    15_00_00 = Column(Integer)
    3PA = Column(Integer)
    3Ppct = Column(Float)
    FTM = Column(Integer)
    FTA = Column(Integer)
    FTpct = Column(Float)
    OREB = Column(Integer)
    DREB = Column(Integer)
    REB = Column(Integer)
    AST = Column(Integer)
    TOV = Column(Integer)
    STL = Column(Integer)
    BLK = Column(Integer)
    PF = Column(Integer)
    FP = Column(Integer)
    DD2 = Column(Integer)
    TD3 = Column(Integer)
    plus_minus = Column(Float)
    OFFRTG = Column(Float)
    DEFRTG = Column(Float)
    NETRTG = Column(Float)
    ASTpct = Column(Float)
    AST/TO = Column(

## Ingestion des Données dans la DB

In [61]:
# j'ai crée la base de donnée avec le fichier create_db.players
# chargons nos différents données
from sqlalchemy.orm import sessionmaker
from sqlalchemy import create_engine

In [5]:
import os, sys
sys.path.append(os.path.abspath('..'))
from utils.create_db import Team, Description, Player, Base


Base SQLite créée avec succès.


In [63]:
engine = create_engine("sqlite:///../data/nba.db")
Session = sessionmaker(bind=engine)
session = Session()

In [64]:
# Insertion teams
for _, row in teams.iterrows():
    session.add(Team(Code=row["Code"], TeamName=row["Nom complet de l'équipe"]))

In [65]:
descriptions = descriptions.astype(str)
# Insertion descriptions
for _, row in descriptions.iterrows():
    session.add(Description(Data=row["Data"], Description=row["Descriptions"]))

In [66]:
# pour le fichier principal convertissons les clés du dataframe
column_mapping = {
    "FG%": "FGpct",
    "3P%": "_3Ppct",
    "3PA": "_3PA",
    "15:00:00": "_15_00_00",
    "+/-": "plus_minus",
    "FT%": "FTpct",
    "AST%": "ASTpct",
    "OREB%": "OREBpct",
    "DREB%": "DREBpct",
    "REB%": "REBpct",
    "EFG%": "EFGpct",
    "TS%": "TSpct",
    "USG%": "USGpct",
    "AST/TO": "AST_TO",
    "AST RATIO": "AST_RATIO",
    "TO RATIO": "TO_RATIO",
}

def convert_keys(data):
    new_data = {}
    for k, v in data.items():
        if k in column_mapping:
            new_data[column_mapping[k]] = v
        else:
            new_data[k] = v
    return new_data

In [67]:
# Inerstion et remplisable des colonnes manquantes
for _, row in players.iterrows(): 
    data = row.to_dict() 
    clean_data = convert_keys(data) 
    session.add(Player(**clean_data))

In [68]:
session.commit() 
session.close()

## NL 2 SQL

In [58]:
from langchain_community.utilities.sql_database import SQLDatabase
from dotenv import load_dotenv
from langchain_mistralai import ChatMistralAI
from langchain.chains.sql_database.query import create_sql_query_chain
from langchain_community.tools.sql_database.tool import QuerySQLDatabaseTool
from langchain_core.prompts import PromptTemplate
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough, RunnableLambda
from operator import itemgetter


In [6]:
db = SQLDatabase.from_uri("sqlite:///../data/nba.db")

In [7]:
print(db.dialect)
print(db.table_info)

sqlite

CREATE TABLE descriptions (
	"Data" VARCHAR NOT NULL, 
	"Description" VARCHAR NOT NULL, 
	PRIMARY KEY ("Data")
)

/*
3 rows from descriptions table:
Data	Description
Player	Nom du joueur
Team	Équipe du joueur (code à 3 lettres)
Age	Âge du joueur
*/


CREATE TABLE players (
	id INTEGER NOT NULL, 
	"Player" VARCHAR, 
	"Team" VARCHAR, 
	"Age" INTEGER, 
	"GP" INTEGER, 
	"W" INTEGER, 
	"L" INTEGER, 
	"Min" FLOAT, 
	"PTS" INTEGER, 
	"FGM" INTEGER, 
	"FGA" INTEGER, 
	"FGpct" FLOAT, 
	_15_00_00 INTEGER, 
	"_3PA" INTEGER, 
	"_3Ppct" FLOAT, 
	"FTM" INTEGER, 
	"FTA" INTEGER, 
	"FTpct" FLOAT, 
	"OREB" INTEGER, 
	"DREB" INTEGER, 
	"REB" INTEGER, 
	"AST" INTEGER, 
	"TOV" INTEGER, 
	"STL" INTEGER, 
	"BLK" INTEGER, 
	"PF" INTEGER, 
	"FP" INTEGER, 
	"DD2" INTEGER, 
	"TD3" INTEGER, 
	plus_minus FLOAT, 
	"OFFRTG" FLOAT, 
	"DEFRTG" FLOAT, 
	"NETRTG" FLOAT, 
	"ASTpct" FLOAT, 
	"AST_TO" FLOAT, 
	"AST_RATIO" FLOAT, 
	"OREBpct" FLOAT, 
	"DREBpct" FLOAT, 
	"REBpct" FLOAT, 
	"TO_RATIO" FLOAT, 
	"EFGpct"

In [8]:
MISTRAL_API_KEY = os.getenv('MISTRAL_API_KEY')

In [9]:
llm = ChatMistralAI(
    mistral_api_key = MISTRAL_API_KEY,
    model='mistral-small-latest',
    temperature=0.1
)
sql_only_prompt = PromptTemplate.from_template("""
Tu es un expert SQL. 
Convertis la question suivante en une requête SQL SQLite valide.
Ne renvoie QUE la requête SQL, sans texte autour.

Tables disponibles :
{table_info}

Nombre maximum de lignes à renvoyer : {top_k}

Question : {input}
SQLQuery:
""")


In [10]:
# notre generateur d'sql
generate_sql = create_sql_query_chain(
    llm=llm,
    db=db,
    prompt=sql_only_prompt
)
sql = generate_sql.invoke({"question": "Il y'a combien de joueur dans ma base de données"})
print(sql)



Parameter `stop` not yet supported (https://docs.mistral.ai/api)


SELECT COUNT(*) FROM players


In [11]:
# Execution de notre sql
execute_query = QuerySQLDatabaseTool(db=db)
execute_query.invoke(sql)

'[(569,)]'

In [12]:
# on peux combiné les deux composants
chain = generate_sql | execute_query  # il prend la sortie du 1er et passe au 2em
chain.invoke({"question": "Il y'a combien de joueur dans ma base de données"})

Parameter `stop` not yet supported (https://docs.mistral.ai/api)


'[(569,)]'

In [13]:
# observons ce qui se passe en interne (le prompt)
chain.get_prompts()[0].pretty_print()


Tu es un expert SQL. 
Convertis la question suivante en une requête SQL SQLite valide.
Ne renvoie QUE la requête SQL, sans texte autour.

Tables disponibles :
{table_info}

Nombre maximum de lignes à renvoyer : 5

Question : {input}
SQLQuery:



In [14]:
# travaillons un peu la sortie de notre système (chain)
answer_prompt = PromptTemplate.from_template(
    """
Tu es un assistant spécialisé en analyse de données NBA.

Ta mission :
1. Lire la question de l'utilisateur.
2. Lire la requête SQL générée.
3. Lire le résultat SQL exécuté.
4. Produire une réponse finale claire, concise et en français.
5. Ne jamais montrer la requête SQL dans la réponse finale.
6. Ne jamais inventer de données : tu dois uniquement interpréter le résultat fourni.

Voici les informations :

Question : {question}
Requête SQL : {query}
Résultat SQL : {result}

Réponse finale :
"""
)



# on passe le prompte au llm
rephrase_answer = answer_prompt | llm | StrOutputParser()

chain = (
    RunnablePassthrough.assign(query=generate_sql)
    .assign(result=itemgetter("query") | execute_query)
    | rephrase_answer
)

In [15]:
chain.invoke({"question": "Il y'a combien de joueur dans ma base de données"})

Parameter `stop` not yet supported (https://docs.mistral.ai/api)


'Il y a **569 joueurs** dans votre base de données.'

### Amélioration de l'agent

In [51]:
# Pour une question de sécurité nous allons ajouté
# Une validation de la requête SQL avant éxecution
def validate_sql(query:str):
    q = query.lower().strip()
    if not q.startswith('select'):
        raise ValueError("Seules les requêtes SELECT sont autorisées.")
    if any(word in q for word in ["drop", "delete", "update", "insert", "alter"]):
        raise ValueError("Requête SQL potentiellement dangereuse.")
    return query

# si aussi la requête SQL échoue
def safe_execute(query: str):
    try:
        return execute_query.run(query)
    except Exception as e:
        return f"ERREUR_SQL: {str(e)}"

#### # Ajoutons des few-shots pour stabiliser la génération SQL

In [39]:
print(db.get_table_info())


CREATE TABLE descriptions (
	"Data" VARCHAR NOT NULL, 
	"Description" VARCHAR NOT NULL, 
	PRIMARY KEY ("Data")
)

/*
3 rows from descriptions table:
Data	Description
Player	Nom du joueur
Team	Équipe du joueur (code à 3 lettres)
Age	Âge du joueur
*/


CREATE TABLE players (
	id INTEGER NOT NULL, 
	"Player" VARCHAR, 
	"Team" VARCHAR, 
	"Age" INTEGER, 
	"GP" INTEGER, 
	"W" INTEGER, 
	"L" INTEGER, 
	"Min" FLOAT, 
	"PTS" INTEGER, 
	"FGM" INTEGER, 
	"FGA" INTEGER, 
	"FGpct" FLOAT, 
	_15_00_00 INTEGER, 
	"_3PA" INTEGER, 
	"_3Ppct" FLOAT, 
	"FTM" INTEGER, 
	"FTA" INTEGER, 
	"FTpct" FLOAT, 
	"OREB" INTEGER, 
	"DREB" INTEGER, 
	"REB" INTEGER, 
	"AST" INTEGER, 
	"TOV" INTEGER, 
	"STL" INTEGER, 
	"BLK" INTEGER, 
	"PF" INTEGER, 
	"FP" INTEGER, 
	"DD2" INTEGER, 
	"TD3" INTEGER, 
	plus_minus FLOAT, 
	"OFFRTG" FLOAT, 
	"DEFRTG" FLOAT, 
	"NETRTG" FLOAT, 
	"ASTpct" FLOAT, 
	"AST_TO" FLOAT, 
	"AST_RATIO" FLOAT, 
	"OREBpct" FLOAT, 
	"DREBpct" FLOAT, 
	"REBpct" FLOAT, 
	"TO_RATIO" FLOAT, 
	"EFGpct" FLOAT,

* **Analyse Rapide :** 
    - Table players : est ma table principale qui est lié à la table teams (players.Team -> teams.Code)
    - Table teams : permet juste de cherché le nom complet de l'équipe
    - Table descriptions : pas utile pour les requêtes mais peux servir pour enrichir les réponses finales

In [41]:
# Constructions de quelques few-shoot à travers la connaissance de notre base de données
few_shoots = """
    Question : Combien de joueurs sont présents dans la base ?
    SQL : SELECT COUNT(*) FROM players;

    Question : Quels sont les joueurs des Denver Nuggets ?
    SQL : SELECT Player FROM players WHERE Team = 'DEN';

    Question : Quelle est la moyenne de points de Nikola Jokić ?
    SQL : SELECT AVG(PTS) FROM players WHERE Player = 'Nikola Jokić';

    Question : Quels sont les 10 meilleurs scoreurs ?
    SQL : SELECT Player, PTS FROM players ORDER BY PTS DESC LIMIT 10;

    Question : Quels joueurs ont plus de 30 ans ?
    SQL : SELECT Player, Age FROM players WHERE Age > 30;

    Question : Quel est le nom complet de l’équipe de Shai Gilgeous-Alexander ?
    SQL : SELECT t.TeamName FROM players p JOIN teams t ON p.Team = t.Code WHERE p.Player = 'Shai Gilgeous-Alexander';

    Question : Trouve les joueurs dont le nom contient "Williams".
    SQL : SELECT Player FROM players WHERE Player LIKE '%Williams%';

    Question : Quels joueurs ont un TSpct supérieur à 60 ?
    SQL : SELECT Player, TSpct FROM players WHERE TSpct > 60;
"""

## Agent final

In [45]:
sql_only_prompt = PromptTemplate.from_template(
f"""
Tu es un expert SQL spécialisé en SQLite et en analyse de données NBA.

Ta mission :
- Lire la question utilisateur.
- Générer une requête SQL SQLite valide.
- Utiliser uniquement les tables et colonnes présentes dans le schéma.
- Ne renvoyer QUE la requête SQL, sans texte autour.

Voici les tables disponibles et leurs schémas :
{{table_info}}

Nombre maximum de lignes à renvoyer : {{top_k}}

Voici quelques exemples (few-shots) pour t’aider à comprendre le style attendu :

{few_shoots}

Règles importantes :
- Utilise uniquement les colonnes présentes dans le schéma.
- Pas de texte hors SQL.
- Pas de commentaires.
- Pas de guillemets inutiles.
- Pas de LIMIT 1 inutile.
- Si la question demande un top N, utilise ORDER BY + LIMIT.
- Si la question mentionne une équipe, utilise la colonne Team (code à 3 lettres).
- Si la question mentionne un joueur, utilise la colonne Player (nom complet exact).

Question utilisateur :
{{input}}

SQLQuery:
"""
)

In [47]:
generate_sql = create_sql_query_chain(
    llm=llm,
    db=db,
    prompt=sql_only_prompt
)
sql = generate_sql.invoke({"question": "Donne-moi les 5 meilleurs rebondeurs"})
print(sql)

Parameter `stop` not yet supported (https://docs.mistral.ai/api)


SELECT Player, REB FROM players ORDER BY REB DESC LIMIT 5;


In [52]:
safe_execute(sql)

"[('Ivica Zubac', 1008), ('Domantas Sabonis', 973), ('Karl-Anthony Towns', 922), ('Nikola Jokić', 889), ('Jalen Duren', 803)]"

In [53]:
# le prompt de sortie final
answer_prompt = PromptTemplate.from_template(
"""
Tu es un assistant expert en analyse de données NBA.

Ton rôle est d’interpréter des résultats SQL et de produire une réponse claire,
naturelle et utile pour l’utilisateur.

### Règles obligatoires :
1. Lis attentivement la question de l’utilisateur.
2. Lis la requête SQL générée (ne jamais la montrer dans la réponse).
3. Lis le résultat SQL exécuté.
4. Réponds uniquement à partir du résultat fourni.
5. Si le résultat est vide, dis-le simplement et propose une reformulation possible.
6. Si le résultat contient une erreur (ex : commence par "ERREUR_SQL"), explique l’erreur
   de manière simple et propose une nouvelle question correcte.
7. La réponse doit être en français, concise, et orientée “insight NBA”.
8. Ne jamais inventer de données. Ne jamais extrapoler au-delà du résultat.

### Informations fournies :

Question utilisateur :
{question}

Requête SQL générée (ne pas afficher dans la réponse) :
{query}

Résultat SQL :
{result}

### Réponse finale (en français, claire, concise, sans SQL) :
"""
)
rephrase_answer = answer_prompt | llm | StrOutputParser()

In [77]:
nl_2_sql = (
    RunnablePassthrough.assign(query=generate_sql) # on genère la requête SQL
    .assign(query=lambda x: validate_sql(x["query"])) # on vérifie la requête SQL
    .assign(result=lambda x: safe_execute(x["query"])) # on exécute la requête
    | rephrase_answer  # puis on reformule la réponse
)


In [79]:
print(nl_2_sql.invoke({"question": "quel est le meilleur joueur et il joue dans quelle équipe"}))

Parameter `stop` not yet supported (https://docs.mistral.ai/api)


Le meilleur joueur selon l'indice de performance est Alondes Williams, qui évolue actuellement avec les Detroit Pistons (DET).
